README.md
curlsoapclient
==============

[![Build Status](https://travis-ci.org/aaharu/curlsoapclient.png?branch=master)](https://travis-ci.org/aaharu/curlsoapclient)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/aaharu/curlsoapclient/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/aaharu/curlsoapclient/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/aaharu/curlsoapclient/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/aaharu/curlsoapclient/?branch=master)

SoapClient with ext-curl.

https://packagist.org/packages/aaharu/curlsoapclient


Documentation
--------------

### Aaharu\Soap\CurlSoapClient::CurlSoapClient

```php
public CurlSoapClient::CurlSoapClient ( mixed $wsdl [, array $options ] )
```

#### supported options

- soap_version
  - either `SOAP_1_1` or `SOAP_1_2`. If omitted, `SOAP_1_1` is used.
- compression
- trace
- exceptions
- connection_timeout
- user_agent
- keep_alive
  - `true` as a default
- login
- password
- proxy_host
- proxy_port
- proxy_login
- proxy_password
- authentication

#### additional options

- redirect_max
  - The maximum amount of HTTP redirections to follow. default is 5.
  - `5` as a default
- curl_timeout
  - CURLOPT_TIMEOUT
  - `30` as a default

#### unsupported options

- stream_context
  - not necessary with curl
- local_cert
  - use `CurlSoapClient::___curlSetOpt` instead
- passphrase
  - use `CurlSoapClient::___curlSetOpt` instead
- ssl_method
  - use `CurlSoapClient::___curlSetOpt` instead


Contributing
--------------

With [composer](https://getcomposer.org) installed, run the following from the root of the repository:

```sh
$ composer install
```

### Running the tests

```sh
$ composer test
```


License
--------------

Licensed under the MIT License.


CHANGELOG
v1.1.0:
  date: 2016-01-16
    changes:
        - fix CurlSoapClient::__getCookies and CurlSoapClient::__setCookie.
        - support proxy_host, proxy_port, proxy_login, proxy_password, login, password and authentication options.
v1.0.0:
  date: 2014-09-25


tests/server.php
<?php
require_once __DIR__ . '/bootstrap.php';

ini_set('date.timezone', 'Asia/Tokyo');
ini_set('soap.wsdl_cache_enabled', 0);

function_exists('getallheaders') && error_log(var_export(getallheaders(), true));

if (@$_GET['auth'] === 'basic') {
    if (!isset($_SERVER['PHP_AUTH_USER']) || $_SERVER['PHP_AUTH_USER'] !== 'hoge') {
        header('WWW-Authenticate: Basic realm="Test Realm"');
        header('HTTP/1.0 401 Unauthorized');
        exit();
    }
} elseif (@$_GET['auth'] === 'digest') {
    $realm = 'Restricted area';
    if (empty($_SERVER['PHP_AUTH_DIGEST'])) {
        header('HTTP/1.1 401 Unauthorized');
        header('WWW-Authenticate: Digest realm="'.$realm.'",qop="auth",nonce="'.uniqid().'",opaque="'.md5($realm).'"');
        exit();
    }
    // through digest validation
}

$redirect = (int) @$_GET['redirect'];
if ($redirect > 0) {
    --$redirect;
    header("Location: http://localhost:8000/tests/server.php?redirect=$redirect");
    exit();
}

if (isset($_GET['503'])) {
    header('HTTP/1.1 503 Service Temporarily Unavailable');
    exit();
}

if (isset($_GET['400'])) {
    header('HTTP/1.1 400 Bad Request');
    header('Content-Type: text/plain');
    echo 'bodybody';
    exit();
}

if (isset($_GET['300'])) {
    header('HTTP/1.1 300');
    exit();
}

$location = @$_GET['location'];
if ($location) {
    header("Location: $location");
    exit();
}

$usleep = (int) @$_GET['usleep'];
if ($usleep > 0) {
    usleep($usleep);
}

function test($x)
{
    return $x;
}

function testFault()
{
    throw new SoapFault('test', 'message');
}

$server = new \SoapServer(null, array('uri' => 'http://test-uri/'));
$server->addFunction('test');
$server->addFunction('testFault');
$server->handle();


tests/CurlSoapClientTest.php
<?php
namespace Aaharu\Soap\CurlSoapClient\Tests;

use Aaharu\Soap\CurlSoapClient;

/**
 * @coversDefaultClass \Aaharu\Soap\CurlSoapClient
 */
class CurlSoapClientTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @test
     */
    public function soap1_1()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?auth=basic',
            'uri' => 'http://test-uri/',
            'compression' => SOAP_COMPRESSION_ACCEPT,
            'connection_timeout' => 1,
            'login' => 'hoge',
            'password' => 'fuga'
        ));

        $response = $obj->test('abc');
        $this->assertEquals('abc', $response);
    }

    /**
     * @test
     */
    public function soap1_2()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?redirect=1&auth=digest',
            'uri' => 'http://test-uri/',
            'user_agent' => 'curlsoapclient',
            'soap_version' => SOAP_1_2,
            'compression' => SOAP_COMPRESSION_GZIP,
            'keep_alive' => false,
            'trace' => true,
            'login' => 'hoge',
            'password' => 'fuga',
            'authentication' => SOAP_AUTHENTICATION_DIGEST
        ));

        $response = $obj->__soapCall('test', array(123));
        $this->assertEquals(123, $response);

        $last_request_headers = $obj->__getLastRequestHeaders();
        $this->assertTrue(stripos($last_request_headers, 'User-Agent: curlsoapclient') !== false);
        $this->assertTrue(stripos($last_request_headers, 'Connection: close') !== false);
    }

    /**
     * @test
     */
    public function overRedirectMax()
    {
        // no exception option
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?redirect=2',
            'uri' => 'http://test-uri/',
            'redirect_max' => 1,
            'exceptions' => false
        ));

        $response = $obj->test(123);
        $this->assertInstanceOf('SoapFault', $response);
        $this->assertTrue(is_soap_fault($response));
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Error Fetching http, 
     */
    public function curlSoapFault()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://noexists',
            'uri' => 'http://test-uri/'
        ));
        $obj->test('hoge');
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Service Temporarily Unavailable
     */
    public function server503()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?503=1',
            'uri' => 'http://test-uri/'
        ));
        $obj->test('hoge');
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage message
     */
    public function testFault()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php',
            'uri' => 'http://test-uri/'
        ));
        $obj->testFault();
    }

    /**
     * @test
     */
    public function http1_0()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php',
            'uri' => 'http://test-uri/',
            'compression' => SOAP_COMPRESSION_DEFLATE,
            'trace' => true
        ));
        $obj->___curlSetOpt(CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_0);
        $class = new \stdClass();
        $response = $obj->test($class);
        $this->assertEquals($class, $response);
        $this->assertTrue(stripos($obj->__getLastRequestHeaders(), 'POST /tests/server.php HTTP/1.0') === 0);
    }

    /**
     * @test
     * @medium
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Error Fetching http, 
     */
    public function timeout()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?usleep=1300000',
            'uri' => 'http://test-uri/',
            'curl_timeout' => 1
        ));
        $class = new \stdClass();
        $obj->test($class);
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Error Redirecting, No Location
     */
    public function noLocation()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?300=1',
            'uri' => 'http://test-uri/'
        ));
        $class = new \stdClass();
        $obj->test($class);
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Error Redirecting, Invalid Location
     */
    public function invalidLocation()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?location=/tmp',
            'uri' => 'http://test-uri/'
        ));
        $obj->test(true);
    }

    /**
     * @test
     */
    public function cookie()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php',
            'uri' => 'http://test-uri/',
            'trace' => true
        ));
        $original_obj = new \SoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php',
            'uri' => 'http://test-uri/',
            'trace' => true
        ));
        $this->assertEquals(array(), $obj->__getCookies());
        $this->assertEquals(array(), $original_obj->__getCookies());
        $obj->__setCookie('CookieTest', 'HelloWorld');
        $obj->__setCookie('CookieTest2', 'HelloWorld2');
        $original_obj->__setCookie('CookieTest', 'HelloWorld');
        $original_obj->__setCookie('CookieTest2', 'HelloWorld2');
        $this->assertEquals(array('CookieTest' => array('HelloWorld'), 'CookieTest2' => array('HelloWorld2')), $obj->__getCookies());
        $this->assertEquals(array('CookieTest' => array('HelloWorld'), 'CookieTest2' => array('HelloWorld2')), $original_obj->__getCookies());
        $this->assertEquals(array(1, 'a', false), $obj->test(array(1, 'a', false)));
        $this->assertEquals(array(1, 'a', false), $original_obj->test(array(1, 'a', false)));
        // difference of CurlSoapClient from SoapClient [";" -> "; "]
        $this->assertTrue(stripos($obj->__getLastRequestHeaders(), 'Cookie: CookieTest=HelloWorld; CookieTest2=HelloWorld2') !== false);
        $this->assertTrue(stripos($original_obj->__getLastRequestHeaders(), 'Cookie: CookieTest=HelloWorld;CookieTest2=HelloWorld2') !== false);
    }

    /**
     * @test
     * @expectedException        \SoapFault
     * @expectedExceptionMessage Bad Request
     */
    public function server400()
    {
        $obj = new CurlSoapClient(null, array(
            'location' => 'http://localhost:8000/tests/server.php?400=1',
            'uri' => 'http://test-uri/'
        ));
        $obj->test(true);
    }
}


src/CurlSoapClient.php
<?php
/**
 * curlsoapclient - SoapClient with ext-curl. -
 *
 * @author    aaharu
 * @copyright Copyright (c) 2014 aaharu
 * @license   MIT License
 */

namespace Aaharu\Soap;

use SoapClient;

/**
 * @see https://github.com/php/php-src/tree/master/ext/soap
 */
class CurlSoapClient extends SoapClient
{
    protected $curl = null; ///< cURL handle
    protected $redirect_max; ///< max redirect counts
    protected $curl_timeout; ///< cURL request time-out seconds
    private $redirect_count = 0;

    public function __construct($wsdl, array $options)
    {
        parent::__construct($wsdl, $options);
        $this->redirect_max = 5;
        if (isset($options['redirect_max'])) {
            $this->redirect_max = (int)$options['redirect_max'];
        }
        $this->curl_timeout = 30;
        if (isset($options['curl_timeout'])) {
            $this->curl_timeout = (int)$options['curl_timeout'];
        }
        $this->curl = curl_init();
        $this->_cookies = array();
    }

    public function __destruct()
    {
        if (isset($this->curl)) {
            curl_close($this->curl);
        }
    }

    public function ___curlSetOpt($option, $value)
    {
        curl_setopt($this->curl, $option, $value);
    }

    public function __getCookies()
    {
        return $this->_cookies;
    }

    public function __setCookie($name, $value = null)
    {
        if (!isset($value)) {
            unset($this->_cookies[$name]);
            return;
        }
        $this->_cookies[$name] = (array)$value;
    }

    /**
     * Execute SOAP requests.
     *
     * @param string $request SOAP request
     * @param string $location SOAP address
     * @param string $action SOAP action
     * @param int $version SOAP version
     * @param int $one_way
     * @throws \Exception
     * @throws \SoapFault
     * @return mixed (string) SOAP response / (object) SoapFault object
     */
    public function __doRequest($request, $location, $action, $version, $one_way = 0)
    {
        curl_setopt($this->curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($this->curl, CURLOPT_HEADER, true);
        curl_setopt($this->curl, CURLOPT_POSTFIELDS, $request);
        if (isset($this->trace) && $this->trace) {
            curl_setopt($this->curl, CURLINFO_HEADER_OUT, true);
        }

        $this->___configHeader($action, $version);
        $this->___configCompression();
        $this->___configTimeout();
        if (isset($this->_user_agent) && is_string($this->_user_agent) && strlen($this->_user_agent) > 0) {
            curl_setopt($this->curl, CURLOPT_USERAGENT, $this->_user_agent);
        }
        $this->___configHttpAuthentication();
        $this->___configProxy();

        try {
            $response = $this->___curlCall($location);
        } catch (\SoapFault $fault) {
            if (isset($this->_exceptions) && empty($this->_exceptions)) {
                // if exceptions option is false, return \SoapFault object
                return $fault;
            }
            throw $fault;
        }

        if ($one_way) {
            return '';
        }

        return $response;
    }

    /**
     * set CURLOPT_HTTPHEADER.
     *
     * @param string $action SOAP action
     * @param int $version SOAP version
     * @return void
     */
    private function ___configHeader($action, $version)
    {
        $header = array();
        if (isset($this->_keep_alive) && empty($this->_keep_alive)) {
            $header[] = 'Connection: close';
        } else {
            $header[] = 'Connection: Keep-Alive';
        }
        if ($version === SOAP_1_2) {
            $header[] = "Content-Type: application/soap+xml; charset=utf-8; action=\"{$action}\"";
        } else {
            $header[] = 'Content-Type: text/xml; charset=utf-8';
            $header[] = "SOAPAction: \"{$action}\"";
        }
        curl_setopt($this->curl, CURLOPT_HTTPHEADER, $header);
    }

    /**
     * set CURLOPT_ENCODING.
     *
     * @return void
     */
    private function ___configCompression()
    {
        if (isset($this->compression)) {
            if ($this->compression & SOAP_COMPRESSION_ACCEPT) {
                curl_setopt($this->curl, CURLOPT_ENCODING, '');
            } elseif ($this->compression & SOAP_COMPRESSION_DEFLATE) {
                curl_setopt($this->curl, CURLOPT_ENCODING, 'deflate');
            } else {
                curl_setopt($this->curl, CURLOPT_ENCODING, 'gzip');
            }
        }
    }

    /**
     * set CURLOPT_CONNECTTIMEOUT and CURLOPT_TIMEOUT.
     *
     * @return void
     */
    private function ___configTimeout()
    {
        $connection_timeout = 10; // default
        if (isset($this->_connection_timeout) && is_int($this->_connection_timeout)) {
            $connection_timeout = $this->_connection_timeout;
        }
        curl_setopt($this->curl, CURLOPT_CONNECTTIMEOUT, $connection_timeout);
        curl_setopt($this->curl, CURLOPT_TIMEOUT, $this->curl_timeout);
    }

    /**
     * set CURLOPT_HTTPAUTH.
     *
     * @return void
     */
    private function ___configHttpAuthentication()
    {
        if (isset($this->_login) && is_string($this->_login) && strlen($this->_login) > 0 &&
            isset($this->_password) && is_string($this->_password) && strlen($this->_password) > 0) {
            curl_setopt($this->curl, CURLOPT_USERPWD, $this->_login . ':' . $this->_password);
            if (property_exists($this, '_digest')) {
                curl_setopt($this->curl, CURLOPT_HTTPAUTH, CURLAUTH_ANYSAFE);
            } else {
                curl_setopt($this->curl, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
            }
        }
    }

    /**
     * set proxy options.
     *
     * @return void
     */
    private function ___configProxy()
    {
        if (isset($this->_proxy_host) && is_string($this->_proxy_host) && strlen($this->_proxy_host) > 0) {
            curl_setopt($this->curl, CURLOPT_PROXY, $this->_proxy_host);
        }
        if (isset($this->_proxy_port) && is_integer($this->_proxy_port)) {
            curl_setopt($this->curl, CURLOPT_PROXYPORT, $this->_proxy_port);
        }
        if (isset($this->_proxy_login) && is_string($this->_proxy_login) && strlen($this->_proxy_login) > 0 &&
            isset($this->_proxy_password) && is_string($this->_proxy_password) && strlen($this->_proxy_password) > 0) {
            curl_setopt($this->curl, CURLOPT_PROXYUSERPWD, $this->_proxy_login . ':' . $this->_proxy_password);
            if (property_exists($this, '_digest')) {
                curl_setopt($this->curl, CURLOPT_PROXYAUTH, CURLAUTH_ANYSAFE);
            } else {
                curl_setopt($this->curl, CURLOPT_PROXYAUTH, CURLAUTH_ANY);
            }
        }
    }

    /**
     * Request cURL.
     *
     * @param[in] string $location SOAP address
     * @throws \SoapFault
     * @return mixed response body
     */
    private function ___curlCall($location)
    {
        curl_setopt($this->curl, CURLOPT_URL, $location);

        if (!empty($this->_cookies)) {
            $cookies = array();
            foreach ($this->_cookies as $cookie_name => $cookie_value) {
                $cookies[] = $cookie_name . '=' . $cookie_value[0];
            }
            curl_setopt($this->curl, CURLOPT_COOKIE, implode('; ', $cookies));
        }

        $response = curl_exec($this->curl);
        if ($response === false) {
            throw new \SoapFault(
                'HTTP',
                'Error Fetching http, ' . curl_error($this->curl) . ' (' . curl_errno($this->curl) . ')'
            );
        }

        $header_size = curl_getinfo($this->curl, CURLINFO_HEADER_SIZE);
        $response_header = substr($response, 0, $header_size);
        $response_body = substr($response, $header_size);
        $http_code = curl_getinfo($this->curl, CURLINFO_HTTP_CODE);

        if (isset($this->trace) && $this->trace) {
            $this->__last_request_headers = curl_getinfo($this->curl, CURLINFO_HEADER_OUT);
            $this->__last_response_headers = $response_header;
        }

        if ($http_code >= 300 && $http_code < 400) {
            $tmp = stristr($response_header, 'Location:');
            $line_end = strpos($tmp, "\n"); // "\r" will be trimmed
            if ($line_end === false) {
                throw new \SoapFault('HTTP', 'Error Redirecting, No Location');
            }
            $new_location = trim(substr($tmp, 9, $line_end - 9));
            $url = parse_url($new_location);
            if ($url === false ||
                empty($url['scheme']) ||
                preg_match('/^https?$/i', $url['scheme']) !== 1
            ) {
                throw new \SoapFault('HTTP', 'Error Redirecting, Invalid Location');
            }
            if (++$this->redirect_count > $this->redirect_max) {
                throw new \SoapFault('HTTP', 'Redirection limit reached, aborting');
            }
            return $this->___curlCall($new_location);
        }

        if ($http_code >= 400) {
            $is_error = false;
            $response_length = strlen($response_body);
            if ($response_length === 0) {
                $is_error = true;
            } elseif ($response_length > 0) {
                $is_xml = false;
                $content_type = curl_getinfo($this->curl, CURLINFO_CONTENT_TYPE);
                if ($content_type !== null) {
                    $separator_position = strpos($content_type, ';');
                    if ($separator_position !== false) {
                        $content_type = substr($content_type, 0, $separator_position);
                    }
                    if ($content_type === 'text/xml' || $content_type === 'application/soap+xml') {
                        $is_xml = true;
                    }
                }
                if (!$is_xml) {
                    $str = ltrim($response_body);
                    if (strncmp($str, '<?xml', 5)) {
                        $is_error = true;
                    }
                }
            }

            if ($is_error) {
                $string_http_code = (string)$http_code;
                $code_position = strpos($response_header, $string_http_code);
                $tmp = substr($response_header, $code_position + strlen($string_http_code));
                $http_message = trim(strstr($tmp, "\n", true));
                throw new \SoapFault('HTTP', $http_message);
            }
        }

        return $response_body;
    }
}
